src/client/storage/uploadChunk/uploadChunk.ts

```diff
@@ -0,0 +1,54 @@
+import {delay, hashFn} from '../../../util';
+import Chunk, {CHUNK_UPLOAD_STATUS} from '../../../db/models/chunk';
+import path from 'path';
+import {promises as fs} from 'fs';
+import {log, UPLOAD_CACHE_PATH, UPLOAD_LOOP_INTERVAL, UPLOAD_RETRY_LIMIT} from '../config';
+
+const uploadChunk = async (data: Buffer) => {
+    const chunkId = hashFn(data).toString('hex');
+    const chunk = await Chunk.findByIdOrCreate(chunkId);
+
+    if (chunk.id !== chunkId) {
+        throw new Error(`Unexpected chunk ids mismatch: ${chunkId}, ${chunk.id}`);
+    }
+
+    const chunkPath = path.join(UPLOAD_CACHE_PATH, `chunk_${chunkId}`);
+    await fs.writeFile(chunkPath, data);
+
+    log.debug({chunkId}, 'Enqueuing chunk for upload');
+    chunk.ul_status = CHUNK_UPLOAD_STATUS.ENQUEUED;
+    await chunk.save();
+
+    let uploaded = false;
+
+    const check = async () => {
+        // TODO: any
+        const updatedChunk: any = await Chunk.find(chunkId);
```

Comment:
I think the chunk structure is more or less stable, let's also create a type for it.

Reply:
Let's also leave it for the next PR. `Chunk` is a model, and writing type for it requires rewriting all our `db` folder in typescript.